Source/path.cpp

```diff
@@ -11,116 +11,143 @@
 #include "objects.h"
 
 namespace devilution {
-
 namespace {
 
+constexpr size_t MaxPathNodes = 300;
+
+struct PathNode {
+	static constexpr uint16_t InvalidIndex = std::numeric_limits<uint16_t>::max();
+
+	int16_t x = 0;
+	int16_t y = 0;
+	uint16_t parentIndex = InvalidIndex;
+	uint16_t childIndices[8] = { InvalidIndex, InvalidIndex, InvalidIndex, InvalidIndex, InvalidIndex, InvalidIndex, InvalidIndex, InvalidIndex };
+	uint16_t nextNodeIndex = InvalidIndex;
+	uint8_t f = 0;
+	uint8_t h = 0;
+	uint8_t g = 0;
+
+	[[nodiscard]] Point position() const
+	{
+		return Point { x, y };
+	}
+
+	void addChild(uint16_t childIndex)
+	{
+		size_t index = 0;
+		for (; index < 8; ++index) {
+			if (childIndices[index] == InvalidIndex)
+				break;
+		}
+		childIndices[index] = childIndex;
+	}

```

Comment:
While it's impossible to hit in the current code as tiles only have 8 neighbours (so this is called a max of 8 times for a given node) this tries to write past the end of the array if all child nodes are full. Would it be worth restructuring the function to set and early return in the loop? e.g.
```suggestion
	void addChild(uint16_t childIndex)
	{
		for (auto &index : childIndices) {
			if (index == InvalidIndex) {
				index = childIndex;
				return;
			}
		}
	}
```

Reply:
We can detect the writes past-the-end with ASAN whereas that version would just do nothing silently. Should I add an assert?

Reply:
an assert sounds good

Reply:
Done